File: 1930-unique-length-3-palindromic-subsequences/1930-unique-length-3-palindromic-subsequences.java
class Solution {
    public int countPalindromicSubsequence(String s) {
         int n = s.length();
        int []first=new int[26];
        int [] last=new int[26];
        Arrays.fill(first,-1);
        Arrays.fill(last,-1);

        for (int i = 0; i < n; i++) {
            if (first[s.charAt(i)-'a'] == -1) first[s.charAt(i)-'a'] = i;
            else last[s.charAt(i)- 'a']= i;
        }
        int ans = 0;
        for (int i = 0; i < 26; i++) {
            if (last[i] !=-1) {
                Set<Character> st=new HashSet<>();
                for (int j = first[i] + 1; j < last[i];j++)
                      st.add(s.charAt(j));
                ans += st.size();
            }
        }
        return ans;
    }
}

File: 1930-unique-length-3-palindromic-subsequences/1930-unique-length-3-palindromic-subsequences.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        int n=s.length();
        vector<pair<int,int>>v(26,{-1,-1});
        for(int i=0;i<n;i++){
            if(v[s[i]-'a'].first==-1) v[s[i]-'a'].first=i;
            else v[s[i]-'a'].second=i;
        }
        int ans=0;
        for(int i=0;i<26;i++){
            if(v[i].second!=-1){
                unordered_set<char>st;
                for(int j=v[i].first+1;j<v[i].second;j++)
                    st.insert(s[j]);
                ans+=st.size();
            }
        }
        return ans;
    }
};